IA-1591 Add GCE+Dataproc algebra to Back Leo

rtitle opened the pull request:
Most of the meat of GCE is in here. High level changs:

* Removed `ClusterHelper`, `ComputeHelper` -- created `RuntimeAlgebra`, `DataprocInterpreter`, `GceInterpreter` which is the main integration point with Google/wb-libs.
* Made `ClusterMonitorActor`/`ClusterMonitorSupervisor` handle both GCE and Dataproc
* Made `LeoPubsubMessageSubscriber` handle both GCE and Dataproc
* Made `BucketHelper`, `VPCHelper` take an `F[_]` type parameter instead of `IO`
* Some pubsub model refactoring/renaming

TODO:
- ~Test the actual GCE creation~
- Add routes for GCE stop/start/delete/etc
- Zombie monitor is still Dataproc specific
- Error handling -- implement GCE getErrorDetails() and clean up on failure
   - In fact GCE doesn't fail if the init script fails, how to signal errors?
- Create dedicated VPC subnet instead of using default
- Merge wb-libs PR once all is working & tested
- Improve unit tests around moved/new files
- (Eventually) add Swagger page for new routes

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

rtitle committed:
Add RuntimeAlgebra and made ClusterHelper extend from it

rtitle committed:
It compiles

rtitle committed:
scalafmt

rtitle committed:
Refactor pubsub models

rtitle committed:
progress

rtitle committed:
Progress, and fixed unit tests

rtitle committed:
Fill in missing GceInterpreter implementations

rtitle committed:
Wire up both algs

rtitle committed:
Fix test compile error

rtitle committed:
Implement ClusterMonitorActor changes

rtitle committed:
Post rebase fixes

rtitle committed:
Bug fix

rtitle commented:
jenkins retest

rtitle committed:
bug fix

rtitle committed:
Fix test

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala:
```diff
@@ -77,7 +79,8 @@ object ClusterMonitorActor {
     authProvider: LeoAuthProvider[IO],
-    clusterHelper: ClusterHelper,
+    dataprocAlg: DataprocAlgebra[IO],
+    gceAlg: RuntimeAlgebra[IO],
```
should it be `GceAlgebra` 🤔 ?

rtitle replied:
So the way I've modeled it is:
- `RuntimeAlgebra` contains all base functionality for runtimes
- `DataprocAlgebra` extends `RuntimeAlgebra` to add Dataproc extensions to the algebra 
- `GceInterpreter` interprets `RuntimeAlgebra`
- `DataprocInterpreter` interprets `DataprocAlgebra`

However there will eventually be GCE extensions to the algebra I think (e.g. for PDs, GPUs, etc) so maybe I will go ahead and make a `GceAlgebra` extending `RuntimeAlgebra`.

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -555,7 +540,7 @@ class LeonardoService(
       case Some(updatedMasterMachineType) if existingCluster.status == Stopped =>
         for {
-          _ <- clusterHelper.updateMasterMachineType(existingCluster, updatedMasterMachineType)
+          _ <- dataprocAlg.updateMachineType(UpdateMachineTypeParams(existingCluster, updatedMasterMachineType))
```
should it be `updateMasterMachineType` still? it seems a bit ambiguous if it's just `updateMachineType`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BaseRuntimeInterpreter.scala:
```diff
@@ -0,0 +1,224 @@
+        _ <- stopGoogleRuntime(params.runtimeAndRuntimeConfig.runtime, params.runtimeAndRuntimeConfig.runtimeConfig)
+
+        // Update the cluster status to Stopping
+        now <- Async[F].delay(Instant.now)
```
shall we have `now` being passed in since it'll exist in api's context?

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala:
```diff
@@ -0,0 +1,644 @@
+
+  import dbRef._
+
+  implicit val contextShiftIO = IO.contextShift(blocker.blockingContext)
```
why this cs is backed by blockingContext?

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/VPCHelper.scala:
```diff
@@ -12,33 +14,34 @@ import org.broadinstitute.dsde.workbench.model.google.GoogleProject
 
-  private def getVPCSettingsFromProjectLabel(googleProject: GoogleProject): IO[Option[VPCConfig]] =
-    IO.fromFuture(IO(googleProjectDAO.getLabels(googleProject.value))).map { labelMap =>
+  implicit val contextShiftIO = IO.contextShift(blocker.blockingContext)
```
same question, why this needs to be backed by a blocking context? also the global `ContextShift[F]` is in scope via contextbound as well

rtitle replied:
Yeah this probably isn't correct.. basically for `IO.fromFuture(…)` I need a `ContextShift[IO]`. But I only have a `ContextShift[F]`. I couldn't figure out how to create a `ContextShift[IO]` from a `ContextShift[F]` so I defined one in this way using the blocking context.

However now I think it might be better for this class to take an implicit `ContextShift[IO]` in its constructor and have `Boot`/`IOApp` provide it.

rtitle committed:
PR feedback so far

rtitle committed:
bugfix

rtitle committed:
GCE fixes

rtitle commented:
jenkins retest

rtitle committed:
Fix templating for gce-init.sh

rtitle committed:
A few more fixes

rtitle committed:
Fix ordering issue

rtitle committed:
More fixes

rtitle committed:
Fixes to the image creation script

rtitle committed:
Tmp branch REVERT ME

rtitle committed:
comment syntax

rtitle committed:
Fix jenkinsfile command

rtitle committed:
Update GCE image with fixes

rtitle committed:
Fix image path

rtitle committed:
Bug fixes, it works now

rtitle committed:
Fix unit test

Qi77Qi commented on http/src/main/resources/init-resources/gce-init.sh:
```diff
@@ -200,7 +200,11 @@ END
-    service google-fluentd reload
+    # restarting instead of `service google-fluentd-reload` because of bug:
+    # https://github.com/GoogleCloudPlatform/google-fluentd/issues/232
+    service google-fluentd restart
```
so fluentd does actually come with gce instances by default?

rtitle replied:
No.. before I installed it I was getting `service not found: google-fluentd`. I think this is a bug in the latest version.

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:
```diff
@@ -140,20 +139,30 @@ class LeoPubsubMessageSubscriber[F[_]: Async: Timer: ContextShift: Concurrent](
                     // perform gddao and db updates for new resources
-                    _ <- Async[F].liftIO(clusterHelper.updateMasterMachineType(resolvedCluster, machineType))
-                    now <- Timer[F].clock.realTime(TimeUnit.MILLISECONDS)
+                    _ <- runtimeConfig.cloudService match {
```
I wonder if it would make sense to have a type class something like 
```
trait RuntimeAlg[F] {
   def updateMachineType(UpdateMachineTypeParams(resolvedCluster, machineType, now)): F[Unit]
   def startRuntime(StartRuntimeParams(resolvedCluster, now)): F[Unit]
/// other common functions
}
```

And then we can do ```runtimeAlg.updateMachineType(xxx)``` etc

rtitle replied:
Trying to understand how a typeclass would work here. With a typeclass I guess I'd summon the instances like this:
```
cloudService match {
  case GCE => RuntimeAlg[F, GCE].updateMachineType(..)
  case Dataproc => RuntimeAlg[F, Dataproc].updateMachineType(..)
}
```
That might be a good way to do it instead of passing each interpreter to the constructor. But I don't see how it replaces the abstract class `BaseRuntimeInterpter` which is used for common functionality between GCE and Dataproc. Am I missing something?

Qi77Qi replied:
maybe not get rid of `BaseRuntimeInterpter `, this is what I was thinking roughly

```
trait RuntimeAlgebra[F[_], A]  {
  def createRuntime(params: CreateRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[CreateRuntimeResponse]
  def deleteRuntime(params: DeleteRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def finalizeDelete(params: FinalizeDeleteParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def stopRuntime(params: StopRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def startRuntime(params: StartRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def updateMachineType(params: UpdateMachineTypeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
  def updateDiskSize(params: UpdateDiskSizeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit]
}

// instances
object RuntimeInstances {
 def gce[F[_]: Async: ContextShift: Logger: DbReference: NewRelicMetrics](config: RuntimeInterpreterConfig.GceInterpreterConfig,
                                            welderDao: WelderDAO[F], ,
                                                                          gceAlgebra: GceAlgebra[F])(implicit ec: ExecutionContext): RuntimeAlgebra[F, CloudService.GCE.type] = new BaseRuntimeInterpreter(config, welderDao){
   override protected def stopGoogleRuntime(runtime: Runtime, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override protected def startGoogleRuntime(runtime: Runtime, welderAction: WelderAction, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override protected def setMachineTypeInGoogle(runtime: Runtime, machineType: MachineTypeName)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override def createRuntime(params: CreateRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[CreateRuntimeResponse] = ???

   override def deleteRuntime(params: DeleteRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override def finalizeDelete(params: FinalizeDeleteParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

   override def updateDiskSize(params: UpdateDiskSizeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???
 }
  
  def dataproc[F[_]: Async: ContextShift: Logger: DbReference: NewRelicMetrics](config: RuntimeInterpreterConfig.DataprocInterpreterConfig,
                                                                                welderDao: WelderDAO[F], ,
                                                                         dataprocAlgebra: dataprocAlgebra[F])(implicit ec: ExecutionContext): RuntimeAlgebra[F, CloudService.Dataproc.type] = new BaseRuntimeInterpreter(config, welderDao){
    override protected def stopGoogleRuntime(runtime: Runtime, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override protected def startGoogleRuntime(runtime: Runtime, welderAction: WelderAction, runtimeConfig: RuntimeConfig)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override protected def setMachineTypeInGoogle(runtime: Runtime, machineType: MachineTypeName)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override def createRuntime(params: CreateRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[CreateRuntimeResponse] = ???

    override def deleteRuntime(params: DeleteRuntimeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override def finalizeDelete(params: FinalizeDeleteParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???

    override def updateDiskSize(params: UpdateDiskSizeParams)(implicit ev: ApplicativeAsk[F, TraceId]): F[Unit] = ???
  }
}
```

rtitle replied:
I think I see.. want to see how the code looks

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BaseRuntimeInterpreter.scala:
```diff
@@ -0,0 +1,230 @@
+import scala.concurrent.ExecutionContext
+import scala.util.Try
+
+abstract private[util] class BaseRuntimeInterpreter[F[_]: Async: ContextShift: Logger](
```
maybe turn this into a typeclass instead? 🤔

rtitle commented:
> Is `sudo` needed?

Nope, fixed :)

Qi77Qi commented on http/src/main/resources/init-resources/jupyter-docker-compose-gce.yaml:
```diff
@@ -14,7 +14,7 @@ services:
     volumes:
       # shared with welder
-      - "${NOTEBOOKS_DIR}":"${NOTEBOOKS_DIR}"
+      - "/work:${NOTEBOOKS_DIR}"
```
is this intentional?

rtitle replied:
Yeah.. I think it was a typo before, we should be mounting `/work` to `${NOTEBOOKS_DIR}`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/HttpGoogleDataprocDAO.scala:
```diff
@@ -412,8 +412,8 @@ class HttpGoogleDataprocDAO(
 
-  private def getOperationUUID(dop: DataprocOperation): UUID =
-    UUID.fromString(dop.getMetadata.get("clusterUuid").toString)
+  private def getOperationUUID(dop: DataprocOperation): String =
```
change this to `private def getGoogleId(dop: DataprocOperation): GoogleId`?

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala:
```diff
@@ -140,7 +150,8 @@ class ClusterMonitorActor(
   val dbRef: DbReference[IO],
   val authProvider: LeoAuthProvider[IO],
-  val clusterHelper: ClusterHelper,
+  val dataprocAlg: DataprocAlgebra[IO],
```
maybe we can just pass in `RuntimeAlgebra[IO]` here? it doesn't seem like there's anything gce or dataproc specific things going on in this class. Do you foresee dataproc or gce specific ops needed in `ClusterMonitorActor`? 🤔

rtitle replied:
So for things like `finalizeDelete` we need to choose the right interpreter depending on whether it's Dataproc or GCE (Dataproc needs to clean up some IAM stuff; GCE doesn't). That's why I'm passing 2 instances here.

How would it look like in the above approach? Maybe:
```
RuntimeInstances.fromCloudService(runtimeConfig.cloudService).finalizeDelete(params)
```
That might be a bit cleaner for callers..

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:
```diff
@@ -1,51 +1,43 @@
   subscriber: GoogleSubscriber[F, LeoPubsubMessage],
-  clusterHelper: ClusterHelper
+  dataprocAlg: DataprocAlgebra[F],
+  gceAlg: GceAlgebra[F]
```
pass in `RuntimeAlgebra` seems will be sufficient for this class?

rtitle replied:
Same as above, maybe we don't pass in any algebras and use `RuntimeInstances.fromCloudService` to retrieve instances?

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala:
```diff
@@ -16,39 +15,39 @@ sealed trait LeoPubsubMessage {
 
 object LeoPubsubMessage {
-  final case class StopUpdate(updatedMachineConfig: RuntimeConfig, clusterId: Long, traceId: Option[TraceId])
+  final case class StopUpdateMessage(updatedMachineConfig: RuntimeConfig, runtimeId: Long, traceId: Option[TraceId])
```
nit and personal preference, the `Message` suffix seems a bit redundant in the context of `LeoPubsubMessage`

rtitle replied:
Hm, I changed it because it wasn't clear to me always that this was a pubsub message since we often `import org.broadinstitute.dsde.workbench.leonardo.monitor.LeoPubsubMessage._`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -635,20 +626,10 @@ class LeonardoService(
-        // Delete the notebook service account key in Google, if present
-        keyIdOpt <- clusterQuery.getServiceAccountKeyId(cluster.googleProject, cluster.runtimeName).transaction
-        _ <- clusterHelper
-          .removeServiceAccountKey(cluster.googleProject, cluster.serviceAccountInfo.notebookServiceAccount, keyIdOpt)
```
is this still needed for dataproc?

rtitle replied:
I moved it to `DataprocInterpreter.deleteRuntime`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -683,30 +664,10 @@ class LeonardoService(
+      _ <- dataprocAlg.stopRuntime(StopRuntimeParams(RuntimeAndRuntimeConfig(cluster, runtimeConfig), now))
     } yield ()
 
-  def internalStopCluster(cluster: Runtime)(implicit ev: ApplicativeAsk[IO, TraceId]): IO[Unit] =
```
👍

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala:
```diff
@@ -721,8 +682,8 @@ class LeonardoService(
 
-      now <- timer.clock.realTime(TimeUnit.MILLISECONDS)
-      _ <- clusterHelper.startCluster(cluster, Instant.ofEpochMilli(now))
+      now <- IO(Instant.now)
```
I think using `Timer` is better than `Instant`. With `Timer`, we can manipulate the time more easily in unit tests if we decided to check time somewhere

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BaseRuntimeInterpreter.scala:
```diff
@@ -0,0 +1,230 @@
+      )
+
+  private def getWelderAction(runtime: Runtime): WelderAction =
+    if (runtime.welderEnabled) {
```
minor, wonder if we can remove `NoAction` case, and have this function return `Option[WelderAction]`. This way, caller can do `getWelderAction(???).traverse(action => ???)`

rtitle replied:
Hm, I guess `NoAction` could be represented as `None`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BaseRuntimeInterpreter.scala:
```diff
@@ -0,0 +1,230 @@
+      .to[Array]
+      .map { bytes =>
+        Map(googleKey -> new String(bytes, StandardCharsets.UTF_8))
+      }
```
this can be 
```
TemplateHelper
      .templateResource[F](replacements, config.clusterResourcesConfig.startupScript, blocker)
      .through(fs2.text.utf8Decode)
      .compile
      .string
      .map { s =>
        Map(googleKey -> s)
      }
```
`string` operator combines `Stream[F, String]` to a single `String`

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/BaseRuntimeInterpreter.scala:
```diff
@@ -0,0 +1,230 @@
+    )
+    val replacements = RuntimeTemplateValues(templateConfig).toMap
+
+    TemplateHelper
```
same comment as above

kyuksel commented on http/src/main/resources/init-resources/jupyter-docker-compose-gce.yaml:
```diff
@@ -14,7 +14,7 @@ services:
     volumes:
       # shared with welder
-      - "${NOTEBOOKS_DIR}":"${NOTEBOOKS_DIR}"
+      - "/work:${NOTEBOOKS_DIR}"
```
Not specific to this PR but a general question: Why are we placing Welder files in the same directory as user files like notebooks? I'd think we'd want to separate files that users may want to view vs. those that are helpful to us for debugging/auditing purposes.

rtitle replied:
Welder needs to have a shared volume with the Jupyter container because of its design -- it syncs notebooks on the filesystem with GCS. It's true that `welder.log` is placed in the same directory but I see that as a benefit as it gives debugging info to the user. I don't think there are any other welder-specific files that are shared (unless I'm forgetting something)..

kyuksel replied:
I can see `welder.log` be useful for power users.

rtitle committed:
PR feedback and cleanup

rtitle committed:
Post rebase fix

rtitle commented:
@Qi77Qi I think I applied all feedback and cleaned stuff up, please take another look.

See especially the `RuntimeInstances` class, I added the syntax:
```
cloudService.interpreter.deleteRuntime(params)
```
(similar to how `DbRef` `.transaction` works)

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala:
```diff
@@ -85,15 +84,17 @@ object RuntimeStatus extends Enum[RuntimeStatus] {
 
   // Non-terminal statuses. Requires monitoring via ClusterMonitorActor.
-  val monitoredStatuses: Set[RuntimeStatus] = Set(Unknown, Creating, Updating, Deleting, Stopping, Starting)
+  val monitoredStatuses: Set[RuntimeStatus] = Set(Unknown, Creating, Updating, Deleting, Stopping, Starting, Staging)
```
we're introducing a new status? can we have newline for each of the status and add some comment about what each status means?

rtitle replied:
Hm, `RuntimeStatus` is kind of overloaded. It's used for both _Leo_ runtime status and _Google_ status for the Dataproc cluster/GCE instance. I added `Staging` because that's a Google status for GCE, but we never actually update the Leo runtime status to `Staging` (we still use `Creating`). So all this change does is prevent `ClusterMonitorActor` from logging "Unknown status" log messages.

I think there's some work to be done to revisit our statuses and monitoring code. I'd prefer to do that in a separate ticket/PR though.

kyuksel replied:
+1 that when we revisit statuses, it'd be clarifying to separate our statuses from Google statuses.

rtitle replied:
Yeah, maybe could be done as part of bullet 2 in https://broadworkbench.atlassian.net/browse/IA-1747

Qi77Qi replied:
yea...I was planning to address this as part of that ticket

kyuksel replied:
👍 Added a note to that ticket.

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala:
```diff
@@ -350,3 +351,19 @@ object VPCConfig {
   final case class VPCSubnet(value: String) extends VPCConfig
 }
+
+sealed trait RuntimeOperation extends Product with Serializable {
```
is this only used for init script?

rtitle replied:
yes.. could move this model to `RuntimeTemplateValues.scala` I guess

rtitle committed:
Post-rebase fixes

rtitle committed:
Don't make the same VPC subnet mistake in GCE
